stanford-earth/SE3_BLT
# SITES-8: Add Acquia Pipelines and CD

# READY FOR REVIEW

## ACTION/DECISION REQUIRED FIRST

_I think we might want to change the Dev environment's branch to pipelines-build-develop, based on the assumption this is the Acquia repository Pipelines mapping to GitHub's develop branch.  For me, it's unclear what the relationship is between develop in GitHub, and develop-build and pipelines-build-develop.  But it sounds from documentation like Pipelines will look among Earth's Environments to see which is running the branch being merged into.  And we'd like it to find the branch currently running on the Dev environment._

# Summary
- This acquia-pipelines.yml file will deploy an Acquia ephemeral environment when a pull request is created in GitHub.  Once it is merged, the file should also remove an ephemeral environment once the pull request is merged or closed.
- After Pipelines completes, a user will still have to manually update the database and files in Acquia's Environment dashboard.  It is unclear to me at the moment how to do this from within Pipelines, once the site is deployed, whatever commands I can run in the Pipelines environment will no longer affect the ODE.  And only code is committed to the ODE, from what I can tell.
- When Earth has reached the maximum number of ODE's allowed on their account, Pipelines will fail with the message, "Unable to create Cloud environment: This application already has the maximum number of CD environments."  A user will have to manually remove an ODE and then re-run the related PR job in order to create an ODE for that new PR.

# Needed By (8/24)
- Today would be great!

# Steps to Test

1. Create a PR in GitHub.
2. Visit the Acquia Pipelines dashboard and you should see the process running.
3. Once complete, copy over the database and files from staging to your new ODE.
4. Visit the ODE URL and you should see the Earth site.

Site created by this PR: http://earthode16.prod.acquia-sites.com

# Associated Issues and/or People
- JIRA ticket: https://stanfordits.atlassian.net/browse/SITES-8

# Extra note
- I chose not to rebase the commits made to this branch, as they contain quite useful information.  During testing, I lost a lot of good information by rebasing too early.
- More detailed information about Pipelines/CD discovery can be found here: https://docs.google.com/document/d/1DAoQHLwqRwjRmY-_oCjhcN1VJAritX-aaqXEK_ooMj8/edit#heading=h.t6v4j4f5kuo6.  Official documentation is both sparse and vague: https://github.com/acquia/pipelines-examples/tree/master/tutorial-701.
